### ISSUE TITLE
Synchronously try upserting webhooks for BBS

### ISSUE THREAD

**Issue by eseliger** [sourcegraph/sourcegraph#10635]
When updating a code host connection, we do checks for the token which then might yield a warning. I think we should do the same for webhooks on BBS, instead of under the hood not telling people about the status.

**Comment by ryanslade** [comment]
I'd like to clarify the flow here before starting work on this:

When a blank secret is added, we fail with a validation error as we do now

When a non blank secret is added and syncing is enabled we attempt to upsert the webhook. 
On success we save the config.
On failure, we show an error and do not save the config

When a non blank secret is added and syncing is disabled, we simply save the config.

When the `plugin.webhooks` section is removed and syncing is enabled we attempt to delete the webhook.
On success we save the config.
On failure, we show an error and do not save the config

When the `plugin.webhooks` section is removed and syncing is disabled we simply save the config.

- [ ] Remove current background syncing code

@eseliger @mrnugget Happy with the above?

**Comment by eseliger** [comment]
Yes 👌

**Comment by mrnugget** [comment]
Yeah, sounds good.

**Comment by ryanslade** [comment]
There's a complication if the user is adding a new external service as we won't have an id yet. We need an id as we append it to the service URL. 

Perhaps another option would be to continue syncing in the background but display a site alert if we fail to sync?

**Comment by mrnugget** [comment]
What if you loosen the restrictions and say that you always save the config, even if syncing the webhook fails, but you also always do it synchronously?

That way you can save a new external service, get the ID, sync in the same request and display an error.

**Comment by ryanslade** [comment]
Another option would be to always save and still display a warning that the webhook wasn't synced instead. Doing it that way would lose our retry logic though

**Comment by eseliger** [comment]
If we do it in a DBTX, the ID would be known prior to commit, no? 🤔 (Not sure if that is simple though)

**Comment by ryanslade** [comment]
> If we do it in a DBTX, the ID would be known prior to commit, no? 🤔 (Not sure if that is simple though)

That could work, I'll see how easy that would be to fit into the current code

**Comment by ryanslade** [comment]
So, it works, but the code needs a bit of cleanup before it'll be ready.

**Comment by ryanslade** [comment]
I discovered another bug while working on this. If you create multiple Bitbucket Server external services with webhooks enabled our auto webhook creation code will always gives them the same id, `sourcegraph-siteid`. We should be using a combination of`SiteID` and `ExternalServiceID`

I'll change that to `sourcegraph-siteID-externalServiceID`

We'll need to clean out the old webhooks on BBS that were defined with 2 part ids so that we don't end up with both 2 part and 3 part webhooks, leading to double delivery for all messages.

There are a few of ways we could do this:

1. Manual step when upgrading to 3.17
2. Auto sync on startup.
3. Just live with double deliveries

2 would need a way to know if we've already done the cleanup so that we don't make unnecessary API calls. Do we already have a place to save this kind of thing? It would probably have to be in the DB because we could have multiple frontend instances.

Do we have a precedence for this sort of thing? How long should cleanup code stay in our code base?

**Comment by mrnugget** [comment]
> If you create multiple Bitbucket Server external services with webhooks enabled our auto webhook creation code will always gives them the same id, `sourcegraph-siteid`

What are the consequences of that? As far as I understand it, the two webhook configurations would overwrite each other with different endpoints/secrets, is that correct?

Another thought I had: do we need to allow multiple webhooks for the same Bitbucket Server instance? Can't we say: "auto creation is only allowed for one webhook"? I know it's a cop out, but I'm wondering whether this isn't the easiest solution for a case that customers likely won't run into.

**Comment by ryanslade** [comment]
> What are the consequences of that? As far as I understand it, the two webhook configurations would overwrite each other with different endpoints/secrets, is that correct?

Yes, we'd only have one entry per sourcegraph instance so all incoming hooks would only match one external service.
 
> Another thought I had: do we need to allow multiple webhooks for the same Bitbucket Server instance? Can't we say: "auto creation is only allowed for one webhook"? I know it's a cop out, but I'm wondering whether this isn't the easiest solution for a case that customers likely won't run into.

We need a webhook per external service so it boils down to how often customers needs to create more than one pointing at the same BBS instance. 

I think doing auto creation on only one could work, but would be confusing for customers that create a second external service.

**Comment by mrnugget** [comment]
> I think doing auto creation on only one could work, but would be confusing for customers that create a second external service.

Even if we validate and throw an error that says "You can only have automatic-creation enabled for one external service"?

**Comment by ryanslade** [comment]
Yes, would could do that.

I think making it work for more than one service isn't actually that much more work than enforcing auto creation for only one instance. We still need to fix the migration work above either way since that bug already exists in 3.16

**Comment by mrnugget** [comment]
> We still need to fix the migration work above either way since that bug already exists in 3.16

Why? If we only allow one external service to have a webhook configuration, why do we need to migrate? (I thought this would save us from doing that, but apparently I misunderstood)

**Comment by ryanslade** [comment]
If a customer is already in the situation where they've set up more than one hook then the last one they added will have been configured on BBS since the last write will "win". We update the syncs ordered by external service ID. 

If we add the new constraint then we'll only sync the first service. I guess you're right, we don't need to do the migration but we would end up changing which external service the one configured hook points at. 

Also, our constraint check would need an edge case to handle where a customer has already set up more than one instance so that saving all but the first one throws an error.

**Comment by ryanslade** [comment]
Closing this as we've decided to remove this feature
